.vscode/tasks.json

```diff
@@ -62,50 +85,116 @@
+            "command": "docker exec -it ${config:container_name} bash -c 'pytest tests/ --tb=short -v --device ${input:model}'",
             "group": {
                 "kind": "build",
                 "isDefault": true
             },
-            "dependsOn": [
-                "make debug"
-            ],
             "problemMatcher": []
         },
         {
-            "label": "run tests",
+            "label": "Run functional tests (with display)",
             "type": "shell",
-            "command": "cd tests && pytest",
+            // Runs functional tests inside the docker container (with Qt display enabled).
+            "command": "docker exec -it ${config:container_name} bash -c 'pytest tests/ --tb=short -v --device ${input:model} --display'",
             "group": {
                 "kind": "build",
                 "isDefault": true
             },
             "problemMatcher": []
         },
         {
-            "label": "kill Speculos",
+            "label": "Install tests requirements",
             "type": "shell",
-            "command": "pkill -f speculos.py",
+            // Installs functional tests python requirements in the docker container. 
+            "command":"docker exec -it -u 0 ${config:container_name} bash -c 'apk add gcc musl-dev python3-dev && pip install -r tests/requirements.txt'",
             "group": {
                 "kind": "build",
                 "isDefault": true
             },
             "problemMatcher": []
+        },
+        {
+            "label": "Load app on device (Linux)",
+            "type": "shell",
+            // Copies the ledger udev rule file to the /etc/udev/rules.d/ directory if it does not exist, then reloads the rules and triggers udev. After that, it executes make load in the container to load the app on a physical device.
+            "command": "if [ ! -f '/etc/udev/rules.d/${config:linux_udev_ledgerblue_rule_file}' ]; then sudo cp .vscode/${config:linux_udev_ledgerblue_rule_file} /etc/udev/rules.d/ && sudo udevadm control --reload-rules && sudo udevadm trigger; fi && docker exec -it ${config:container_name} bash -c 'export BOLOS_SDK=$(echo ${input:sdk}) && make load'",
+            "group": {
+                "kind": "build",
+                "isDefault": true
+            },
+            "problemMatcher": []
+        },
+        {
+            "label": "Load app on device (macOS)",
+            "type": "shell",
+            // Checks that virtual env is installed, otherwise installs it. Then installs ledgerblue in a virtualenv and side loads the app APDU file.
+            "command": "cmd=\"[ -d 'ledger' ] || python3 -m virtualenv ledger && source ledger/bin/activate && python3 -m pip install ledgerblue && python3 -m ledgerblue.runScript --scp --fileName bin/app.apdu --elfFile bin/app.elf\"; if python3 -m virtualenv --version >/dev/null 2>&1; then eval \"${cmd}\"; else python3 -m pip install virtualenv && eval \"${cmd}\"; fi",
```

Comment:
Maybe there could be also a pre-requirement on virtualenv installation and ledger virtualenv creation, so that here you just activate and launch the command?
And you don't need udev rules on Mac?

Reply:
* I wanted to avoid having too many tasks...
* udev does not exist on macOS. I don't know exactly how low level device stuff are managed, but macOs has no udev rules.

Reply:
Done.